### PR TITLE
Herder: cleanup trigger ledger mechanism

### LINKED ISSUE #3111: Herder: improve ledger trigger mechanism
Currently, Herder uses a combination of tracking and LedgerManager state checks to decide if it needs to trigger next ledger, but the code is very error-prone and hard to follow (and buggy in certain edge cases). It would be much easier to reason about this logic if instead we had a signal from LedgerManager to trigger next ledger when a new ledger is closed.

### PR DESCRIPTION
**Pull request by marta-lokhova**
Resolves #3111

### CODE REVIEW

File: src/herder/HerderImpl.cpp [hunk 1]
```diff
@@ -918,15 +918,11 @@ HerderImpl::eraseBelow(uint32 ledgerSeq)
 }
 
 void
-HerderImpl::ledgerClosed()
+HerderImpl::cleanup()
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderImpl.cpp:932]: Also, I don't really like that we're creating more mini functions that need to be called exactly right: I am not sure why there is drift between what happens in `bootstrap` and the "externalize" path.

I think we should be instead consolidating/reorder things a bit such that we can have 1 function that performs something like:
```
        // start timing next externalize from this point
        mLastExternalize = mApp.getClock().now();

        // perform cleanups
        TxSetFramePtr externalizedSet =
            mPendingEnvelopes.getTxSet(value.txSetHash);
        updateTransactionQueue(externalizedSet->mTransactions);

        cleanup(); // replace by actual content 
        safelyProcessSCPQueueUpToIndex(nextConsensusLedgerIndex(), false);  // maybe replace by actual content?
```

File: src/herder/test/HerderTests.cpp [hunk 2]
```diff
@@ -2016,10 +2016,15 @@ TEST_CASE("herder externalizes values", "[herder]")
             if (i == ledgers.size() - 1)
             {
                 checkSynced(*(getC()));
+                // All the buffered ledgers are applied by now, so it's safe to
+                // trigger the next ledger
+                REQUIRE(herderC.mTriggerNextLedgerSeq == fourth + 1);
```
Comment (MonsieurNicolas) [inline src/herder/test/HerderTests.cpp:2021]: this check is incomplete: what you mean here is that there is a "trigger timer" setup in the future and that it's "the right ledger" (same applies to other places where you check `mTriggerNextLedgerSeq` directly)

File: src/herder/HerderImpl.cpp [hunk 3]
```diff
@@ -844,6 +836,17 @@ HerderImpl::safelyProcessSCPQueueUpToIndex(uint64 nextIndex, bool synchronous)
     }
 }
 
+void
+HerderImpl::lastClosedLedgerIncreased()
+{
+    releaseAssert(isTracking());
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:844]: note that in the context of this PR this is fine, but those invariants may not hold in the future:
for example, if we want to be able to "trigger" at the end of catchup without having to wait for an extra ledger.

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
looks pretty close to me. I made a few suggestions to improve clarity

File: src/herder/test/HerderTests.cpp (hunk 2 above)
Comment (marta-lokhova) [reply src/herder/test/HerderTests.cpp:2023]: I added a check for trigger timer setup. I left the section `older ledgers externalize and no-op` as-is, since it's supposed to skip triggering an older ledger, and the application may or may not have a trigger setup for the most recent ledger.

**Comment by marta-lokhova** [comment]
addressed your comments @MonsieurNicolas in the most recent commit. I'll squash when we're ready to merge

**Comment by MonsieurNicolas** [comment]
looks good. Please rebase @marta-lokhova

**Comment by marta-lokhova** [comment]
@MonsieurNicolas done!

**Comment by MonsieurNicolas** [comment]
r+ 21433d8b7320e3a0c4e5c57023cf3e7194c715ab
